capsules/src/virtual_alarm.rs

```diff
@@ -143,31 +139,39 @@ impl<'a, Alrm: Alarm> time::Client for MuxAlarm<'a, Alrm> {
 
         // Check whether to fire each alarm. At this level, alarms are one-shot,
         // so a repeating client will set it again in the fired() callback.
-        for cur in self.virtual_alarms.iter() {
-            let should_fire = past_from_base(cur.when.get(), now + 100, prev);
-            if cur.armed.get() && should_fire {
+        self.virtual_alarms
+            .iter()
+            .filter(|cur| cur.armed.get() && past_from_base(cur.when.get(), now, prev))
+            .for_each(|cur| {
                 cur.armed.set(false);
                 self.enabled.set(self.enabled.get() - 1);
                 cur.fired();
-            }
-        }
+            });
 
         // Find the soonest alarm client (if any) and set the "next" underlying
-        // alarm based on it.
-        let mut next = None;
-        let mut min_distance: u32 = u32::max_value();
-        for cur in self.virtual_alarms.iter() {
-            if cur.armed.get() {
+        // alarm based on it.  This needs to happen after firing all expired
+        // alarms since those may have reset new alarms.
+        let (next, _) = self.virtual_alarms
+            .iter()
+            .filter(|cur| cur.armed.get())
+            .fold((None, u32::max_value()), |(next, min_distance), cur| {
                 let distance = cur.when.get().wrapping_sub(now);
-                if cur.armed.get() && distance < min_distance {
-                    min_distance = distance;
-                    next = Some(cur);
+                if distance < min_distance {
+                    (Some(cur), distance)
+                } else {
+                    (next, min_distance)
```

Comment:
Isn't a little bit odd use `fold` to replace the value i.e, not used an `accumulator`? Or am I misunderstanding something here?

I guess the last value consumed by `iter` will be next?